File: java/senscord/src/senscord_properties/InterlaceProperty.java
/*
 * SPDX-FileCopyrightText: 2019 Sony Semiconductor Solutions Corporation
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package senscord_properties;

import com.sun.jna.Structure;
import java.util.Arrays;
import java.util.List;

/**
 * Channel's property for interlace.
 * 
 * @see "senscord::InterlaceProperty"
 */
public class InterlaceProperty extends Structure {
    public int field;

    public static class ByReference extends InterlaceProperty implements Structure.ByReference {
    }

    @Override
    protected List<String> getFieldOrder() {
        return Arrays.asList("field");
    }

    /**
     * The field types of interlace.
     * 
     * @see "senscord::InterlaceField"
     */
    public static interface Feild {
        /**
         * Top field.
         */
        public static final int SENSCORD_INTERLACE_FIELD_TOP = 0;

        /**
         * Bottom field.
         */
        public static final int SENSCORD_INTERLACE_FIELD_BOTTOM = 1;

    }

    /**
     * Get contained field type.
     * 
     * @return contained field type
     */
    public float GetField() {
        return field;
    }

    /**
     * Set contained field type.
     * 
     * @param field contained field type
     */
    public void SetField(int field) {
        this.field = field;
    }
}

File: csharp/lib/Property/InterlaceProperty.cs
/*
 * SPDX-FileCopyrightText: 2017-2023 Sony Semiconductor Solutions Corporation
 *
 * SPDX-License-Identifier: Apache-2.0
 */

using MessagePack;

namespace SensCord
{
    /// <summary>
    /// Channel's property for interlace.
    /// </summary>
    [MessagePackObject]
    public class InterlaceProperty : IBaseProperty
    {
        /// <summary>
        /// Property key.
        /// </summary>
        public static readonly string ConstKey = "interlace_property";

        /// <summary>
        /// Property key.
        /// </summary>
        [IgnoreMember]
        public string Key { get; } = ConstKey;

        /// <summary>
        /// contained field type.
        /// </summary>
        [Key("field")]
        public InterlaceField Field { get; set; }

        /// <summary>
        /// To string.
        /// </summary>
        /// <returns>A string that represents the current object.</returns>
        public override string ToString()
        {
            return $"{this.Key}({this.Field})";
        }
    }

    /// <summary>
    /// Property for interlace information.
    /// </summary>
    [MessagePackObject]
    public class InterlaceInfoProperty : IBaseProperty
    {
        /// <summary>
        /// Property key.
        /// </summary>
        public static readonly string ConstKey = "interlace_info_property";

        /// <summary>
        /// Property key.
        /// </summary>
        [IgnoreMember]
        public string Key { get; } = ConstKey;

        /// <summary>
        /// order of field.
        /// </summary>
        [Key("order")]
        public InterlaceOrder Order { get; set; }

        /// <summary>
        /// To string.
        /// </summary>
        /// <returns>A string that represents the current object.</returns>
        public override string ToString()
        {
            return $"{this.Key}({this.Order})";
        }
    }

    /// <summary>
    /// The field types of interlace.
    /// </summary>
    public enum InterlaceField
    {
        /// <summary>
        /// Top field.
        /// </summary>
        Top,

        /// <summary>
        /// Bottom field.
        /// </summary>
        Bottom,
    }

    /// <summary>
    /// The order of interlace.
    /// </summary>
    public enum InterlaceOrder
    {
        /// <summary>
        /// Top first.
        /// </summary>
        TopFirst,

        /// <summary>
        /// Bottom first.
        /// </summary>
        BottomFirst,
    }
}
